File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    int visitedStart = 100000 + 1;
    public ListNode detectCycle(ListNode head) {
        ListNode dummy = new ListNode(-100000 - 1);
        dummy.next = head;
        int visitedTemp = visitedStart + 1;
        while(head != null){
            if(head.val >= visitedStart) {
                return head;
            }
            head.val = visitedTemp++;
            head = head.next;
        }
        return null;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.ts
function detectCycle(head: ListNode | null): ListNode | null {
    let slow: ListNode | null = head;
    let fast: ListNode | null = head;

    // 사이클 탐색
    while (fast !== null && fast.next !== null) {
        slow = slow!.next;
        fast = fast.next.next;

        // 두 포인터가 만나면 사이클 존재
        if (slow === fast) {
            // slow를 처음으로 이동
            slow = head;

            // 사이클 시작점 탐색
            while (slow !== fast) {
                slow = slow!.next;
                fast = fast!.next;
            }

            return slow;
        }
    }

    // 사이클이 없는 경우
    return null;
}